server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MigratingMirrorToRepositoryService.java

```diff
@@ -0,0 +1,195 @@
+        }
+
+        // Exit read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.WRITABLE))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Mirrors to each repository migration has been completed. (took: {} ms.)",
+                    stopwatch.elapsed().toMillis());
+    }
+
+    private void logMigrationJob(int numMigratedProjects) throws Exception {
+        final Map<String, Object> data =
+                ImmutableMap.of("timestamp", Instant.now(),
+                                "projects", numMigratedProjects);
+        final Change<JsonNode> change = Change.ofJsonUpsert(MIRROR_TO_REPOSITORY_MIGRATION_JOB_LOG,
+                                                            Jackson.writeValueAsString(data));
+        final Command<CommitResult> command =
+                Command.push(Author.SYSTEM, InternalProjectInitializer.INTERNAL_PROJECT_DOGMA,
+                             Project.REPO_DOGMA, Revision.HEAD,
+                             "Migration of mirrors to each repository has been done.", "",
+                             Markup.PLAINTEXT, change);
+        executeCommand(command);
+    }
+
+    private boolean hasMigrationLog() throws Exception {
+        final Project internalProj = projectManager.get(InternalProjectInitializer.INTERNAL_PROJECT_DOGMA);
+        final Repository repository = internalProj.repos().get(Project.REPO_DOGMA);
+        final Map<String, Entry<?>> entries = repository.find(Revision.HEAD,
+                                                              MIRROR_TO_REPOSITORY_MIGRATION_JOB_LOG).get();
+        final Entry<?> entry = entries.get(MIRROR_TO_REPOSITORY_MIGRATION_JOB_LOG);
+        return entry != null;
+    }
+
+    private boolean migrateMirrors(MetaRepository repository) throws Exception {
+        final Map<String, Entry<?>> entries = repository.find(Revision.HEAD, LEGACY_MIRRORS_PATH + "*.json")
+                                                        .join();
+        if (entries.isEmpty()) {
+            return false;
+        }
+        final List<Change<?>> changes = new ArrayList<>();
+        for (Map.Entry<String, Entry<?>> entry : entries.entrySet()) {
+            final JsonNode content = (JsonNode) entry.getValue().content();
+            if (!content.isObject()) {
+                warnInvalidMirrorConfig(entry, content);
+                continue;
+            }
+            try {
+                final MirrorConfig mirrorConfig = Jackson.treeToValue(content, MirrorConfig.class);
+                final String repoName = mirrorConfig.localRepo();
+                if (repoName == null) {
+                    warnInvalidMirrorConfig(entry, content);
+                    continue;
+                }
+                final MirrorConfig newMirrorConfig = mirrorConfig.withCredentialId(
+                        projectMirrorCredentialId(repository.parent().name(), mirrorConfig.credentialId()));
```

Comment:
Should we skip migration for a `MirrorConfig` whose the credential ID starts with `projects/`? Because we don't rollback if the migration process is halt somehow. This migration could be partially applied.